packages/backend/src/engine/custom_outputs_accumulator.ts

```diff
@@ -0,0 +1,66 @@
+import { BackendOutputStorageStrategy } from '@aws-amplify/plugin-types';
+import { ClientConfig } from '@aws-amplify/client-config';
+import {
+  CustomOutput,
+  customOutputKey,
+} from '@aws-amplify/backend-output-schemas';
+import { Lazy } from 'aws-cdk-lib';
+import {
+  AmplifyUserError,
+  ObjectAccumulator,
+  ObjectAccumulatorPropertyAlreadyExistsError,
+} from '@aws-amplify/platform-core';
+
+/**
+ * Accumulates custom outputs as they're added to the backend.
+ */
+export class CustomOutputsAccumulator {
+  private hasBackendOutputEntry = false;
```

Comment:
I created this class to offload logic from `BackendFactory` (aka `defineBackend` call).
https://github.com/aws-amplify/amplify-backend/blob/5aaba0982b9311c9d2c511130700bb5c760205e7/packages/backend/src/backend_factory.ts#L65-L68

So, the lifecycle scope of this is 1:1 to `BackendFactory` and therefore 1:1 with "main stack".
Technically there's no limitation that `defineBackend` could be called twice and deploy two main stacks.

Otherwise, we've learnt that static classes sooner or later lead to refactors where we have to make breaking changes to un-static them. We've seen that with Printer and deployer factory recently.
Therefore, we'd rather make classes un-aware of their scope and control the scope via DI "container" (our makeshift container) or by declaring static const that has reference.

Reply:
https://github.com/aws-amplify/amplify-backend/issues/953

Reply:
also, recent printer change for reference https://github.com/aws-amplify/amplify-backend/pull/903